lib/smf/smf.h

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2021 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* State Machine Framework */
+
+#ifndef _SMF_H__
+#define _SMF_H__
+
+#ifndef CONFIG_SMF_FLAT
+/**
+ * @brief Macro to create a hierarchical state that prints a message
+ *        on entry to and exit from the state.
+ *
+ * @param _entry_msg Message printed on state entry
+ * @param _exit_msg  Message printed on state exit
+ * @param _entry     State entry function
+ * @param _run       State run function
+ * @param _exit      State exit function
+ * @param _parent    State parent object or NULL
+ */
+#define HSM_CREATE_MSG_STATE( \
+	_entry_msg, _exit_msg, _entry, _run, _exit, _parent) \
+{ \
+	.entry_msg = _entry_msg, \
+	.exit_msg  = _exit_msg,  \
+	.entry     = _entry,     \
+	.run       = _run,       \
+	.exit      = _exit,      \
+	.parent    = _parent     \
+}
+
+/**
+ * @brief Macro to create a hierarchical state.
+ *
+ * @param _entry  State entry function
+ * @param _run    State run function
+ * @param _exit   State exit function
+ * @param _parent State parent object or NULL
+ */
+#define HSM_CREATE_STATE(_entry, _run, _exit, _parent) \
+{ \
+	.entry  = _entry, \
+	.run    = _run,   \
+	.exit   = _exit,  \
+	.parent = _parent \
+}
+
+#else
+
+/**
+ * @brief Macro to create a flat state that prints a message
+ *        on entry to and exit from the state.
+ *
+ * @param _entry_msg Message printed on state entry
+ * @param _exit_msg  Message printed on state exit
+ * @param _entry     State entry function
+ * @param _run       State run function
+ * @param _exit      State exit function
+ */
+#define FSM_CREATE_MSG_STATE(_entry_msg, _exit_msg, _entry, _run, _exit) \
+{ \
+	.entry_msg = _entry_msg, \
+	.exit_msg  = _exit_msg,  \
+	.entry     = _entry,     \
+	.run       = _run,       \
+	.exit      = _exit       \
+}
```

Comment:
I think you'll need to wrap this macro with CONFIG_SMF_PRINT_MSG so that users can cleanly disable the entry/exit messages with just a config change:

```
#ifdef CONFIG_SMF_PRINT_MSG
#define FSM_CREATE_MSG_STATE() \
   { .entry_msg = _entry_msg, .exit_msg = _exit_msg, .entry = _entry, /* etc */ }
#else
#define FSM_CREATE_MSG_STATE() \
   { .entry = _entry, /* etc */ }
#endif
```

Reply:
I've removed all the code related to PRINT_MSG in favor of using the logging subsystem